### PR TITLE
kmsv2: add apiserver identity to metrics

### LINKED ISSUE #117726: [KMSv2] Add apiserver identity to the metrics
/sig auth
/triage accepted
/assign

### PR DESCRIPTION
**Pull request by ritazh**
/kind feature
/triage accepted
/milestone v1.29

Fixes #117726

#### Does this PR introduce a user-facing change?

```release-note
Adding apiserver identity to the following metrics: 
apiserver_envelope_encryption_key_id_hash_total, apiserver_envelope_encryption_key_id_hash_last_timestamp_seconds, apiserver_envelope_encryption_key_id_hash_status_last_timestamp_seconds, apiserver_encryption_config_controller_automatic_reload_failures_total, apiserver_encryption_config_controller_automatic_reload_success_total, apiserver_encryption_config_controller_automatic_reload_last_timestamp_seconds

Fix bug to surface events for the following metrics: apiserver_encryption_config_controller_automatic_reload_failures_total, apiserver_encryption_config_controller_automatic_reload_last_timestamp_seconds, apiserver_encryption_config_controller_automatic_reload_success_total
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP] https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 1]
```diff
@@ -130,7 +131,7 @@ var (
 			Help:           "The last time in seconds when a keyID was used.",
 			StabilityLevel: metrics.ALPHA,
 		},
-		[]string{"transformation_type", "provider_name", "key_id_hash"},
+		[]string{"transformation_type", "provider_name", "key_id_hash", "apiserver_id"},
```
Comment (ritazh) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:134]: starting with these two metrics, any other metrics need the apiserver identity?

Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:134]: You want all 3 metrics that include the `key_id_hash` label.

You also want all 3 of the `encryption_config_controller` metrics.

Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:134]:
```suggestion
		[]string{"transformation_type", "provider_name", "key_id_hash", "apiserver_id_hash"},
```

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 2]
```diff
@@ -171,13 +171,15 @@ type EncryptionConfiguration struct {
 	// KMSCloseGracePeriod is the duration we will wait before closing old transformers.
 	// We wait for any in-flight requests to finish by using the duration which is longer than their timeout.
 	KMSCloseGracePeriod time.Duration
+
+	APIServerID string
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:175]: I would not expect `LoadEncryptionConfig` to return this.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 3]
```diff
@@ -294,7 +294,7 @@ func (s *EtcdOptions) maybeApplyResourceTransformers(c *server.Config) (err erro
 		}
 	}()
 
-	encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(ctxTransformers, s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload)
+	encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(ctxTransformers, s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, c.APIServerID)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:297]: Pretty sure you need to wire this to `NewDynamicEncryptionConfiguration` as well.

**Review by enj — CHANGES_REQUESTED** [review]

**Comment by mengjiao-liu** [comment]
/sig instrumentation

/retest
for pending CI jobs

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

(for sig inst)

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go [hunk 4]
```diff
@@ -29,58 +32,83 @@ const (
...
 }
 
-func RecordEncryptionConfigAutomaticReloadSuccess() {
-	encryptionConfigAutomaticReloadSuccessTotal.Inc()
-	recordEncryptionConfigAutomaticReloadTimestamp("success")
+func RecordEncryptionConfigAutomaticReloadSuccess(apiServerID string) {
+	apiServerIDHash := ""
+	if len(apiServerID) > 0 {
+		apiServerIDHash = getHash(apiServerID)
+	}
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:98]: Could we make this a function so we always have consistent logic across all calls?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go [hunk 5]
```diff
@@ -29,58 +32,83 @@ const (
...
+	encryptionConfigAutomaticReloadSuccessTotal.WithLabelValues(apiServerIDHash).Inc()
+	recordEncryptionConfigAutomaticReloadTimestamp("success", apiServerIDHash)
 }
 
-func recordEncryptionConfigAutomaticReloadTimestamp(result string) {
-	encryptionConfigAutomaticReloadLastTimestampSeconds.WithLabelValues(result).SetToCurrentTime()
+func recordEncryptionConfigAutomaticReloadTimestamp(result string, apiServerIDHash string) {
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:103]:
```suggestion
func recordEncryptionConfigAutomaticReloadTimestamp(result, apiServerIDHash string) {
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 6]
```diff
@@ -182,19 +183,19 @@ func registerLRUMetrics() {
 
 	keyIDHashTotalMetricLabels = lru.NewWithEvictionFunc(cacheSize, func(key lru.Key, _ interface{}) {
 		item := key.(metricLabels)
-		if deleted := KeyIDHashTotal.DeleteLabelValues(item.transformationType, item.providerName, item.keyIDHash); deleted {
+		if deleted := KeyIDHashTotal.DeleteLabelValues(item.transformationType, item.providerName, item.keyIDHash, item.apiServerIDHash); deleted {
 			klog.InfoS("Deleted keyIDHashTotalMetricLabels", "transformationType", item.transformationType,
-				"providerName", item.providerName, "keyIDHash", item.keyIDHash)
+				"providerName", item.providerName, "keyIDHash", item.keyIDHash, "apiServerID", item.apiServerIDHash)
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:188]:
```suggestion
				"providerName", item.providerName, "keyIDHash", item.keyIDHash, "apiServerIDHash", item.apiServerIDHash)
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 7]
```diff
@@ -182,19 +183,19 @@ func registerLRUMetrics() {
...
 		}
-		if deleted := KeyIDHashLastTimestampSeconds.DeleteLabelValues(item.transformationType, item.providerName, item.keyIDHash); deleted {
+		if deleted := KeyIDHashLastTimestampSeconds.DeleteLabelValues(item.transformationType, item.providerName, item.keyIDHash, item.apiServerIDHash); deleted {
 			klog.InfoS("Deleted keyIDHashLastTimestampSecondsMetricLabels", "transformationType", item.transformationType,
-				"providerName", item.providerName, "keyIDHash", item.keyIDHash)
+				"providerName", item.providerName, "keyIDHash", item.keyIDHash, "apiServerID", item.apiServerIDHash)
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:192]:
```suggestion
				"providerName", item.providerName, "keyIDHash", item.keyIDHash, "apiServerIDHash", item.apiServerIDHash)
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 8]
```diff
@@ -182,19 +183,19 @@ func registerLRUMetrics() {
...
 		}
 	})
 	keyIDHashStatusLastTimestampSecondsMetricLabels = lru.NewWithEvictionFunc(cacheSize, func(key lru.Key, _ interface{}) {
 		item := key.(metricLabels)
-		if deleted := KeyIDHashStatusLastTimestampSeconds.DeleteLabelValues(item.providerName, item.keyIDHash); deleted {
-			klog.InfoS("Deleted keyIDHashStatusLastTimestampSecondsMetricLabels", "providerName", item.providerName, "keyIDHash", item.keyIDHash)
+		if deleted := KeyIDHashStatusLastTimestampSeconds.DeleteLabelValues(item.providerName, item.keyIDHash, item.apiServerIDHash); deleted {
+			klog.InfoS("Deleted keyIDHashStatusLastTimestampSecondsMetricLabels", "providerName", item.providerName, "keyIDHash", item.keyIDHash, "apiServerID", item.apiServerIDHash)
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:198]:
```suggestion
			klog.InfoS("Deleted keyIDHashStatusLastTimestampSecondsMetricLabels", "providerName", item.providerName, "keyIDHash", item.keyIDHash, "apiServerIDHash", item.apiServerIDHash)
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 9]
```diff
@@ -218,22 +219,22 @@ func RegisterMetrics() {
 }
 
 // RecordKeyID records total count and last time in seconds when a KeyID was used for TransformFromStorage and TransformToStorage operations
-func RecordKeyID(transformationType, providerName, keyID string) {
+func RecordKeyID(transformationType, providerName, keyID string, apiServerID string) {
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:222]:
```suggestion
func RecordKeyID(transformationType, providerName, keyID, apiServerID string) {
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 10]
```diff
@@ -218,22 +219,22 @@ func RegisterMetrics() {
...
-	KeyIDHashTotal.WithLabelValues(transformationType, providerName, keyIDHash).Inc()
-	KeyIDHashLastTimestampSeconds.WithLabelValues(transformationType, providerName, keyIDHash).SetToCurrentTime()
+	keyIDHash, apiServerIDHash := addLabelToCache(keyIDHashTotalMetricLabels, transformationType, providerName, keyID, apiServerID)
+	KeyIDHashTotal.WithLabelValues(transformationType, providerName, keyIDHash, apiServerIDHash).Inc()
+	KeyIDHashLastTimestampSeconds.WithLabelValues(transformationType, providerName, keyIDHash, apiServerIDHash).SetToCurrentTime()
 }
 
 // RecordKeyIDFromStatus records last time in seconds when a KeyID was returned by the Status RPC call.
-func RecordKeyIDFromStatus(providerName, keyID string) {
+func RecordKeyIDFromStatus(providerName, keyID string, apiServerID string) {
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:232]:
```suggestion
func RecordKeyIDFromStatus(providerName, keyID, apiServerID string) {
```

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go [hunk 11]
```diff
@@ -305,16 +306,21 @@ func getHash(data string) string {
 	return result
 }
 
-func addLabelToCache(c *lru.Cache, transformationType, providerName, keyID string) string {
+func addLabelToCache(c *lru.Cache, transformationType, providerName, keyID string, apiServerID string) (string, string) {
```
Comment (aramase) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go:309]:
```suggestion
func addLabelToCache(c *lru.Cache, transformationType, providerName, keyID, apiServerID string) (string, string) {
```

**Review by enj — CHANGES_REQUESTED** [review]
I would expect to see this new label in our KMS v2 integration tests (you will need to use the `SetHostnameFuncForTests` to get a stable hash value). 

Also, I see a segfault in the CI run which needs to be fixed:

```
I0908 04:36:52.172784   87340 kms_plugin_mock.go:114] "Starting KMS Plugin" socketPath="@new-kms-provider.sock"
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x47f97d]

goroutine 10621 [running]:
runtime.throw({0x4b93a49?, 0xc002797ac8?})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/.gimme/versions/go1.21.1.linux.amd64/src/runtime/panic.go:1077 +0x5c fp=0xc002797a88 sp=0xc002797a58 pc=0x43d87c
runtime.sigpanic()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/.gimme/versions/go1.21.1.linux.amd64/src/runtime/signal_unix.go:845 +0x3e9 fp=0xc002797ae8 sp=0xc002797a88 pc=0x454b69
sync.(*Pool).pin(0x0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/.gimme/versions/go1.21.1.linux.amd64/src/sync/pool.go:204 +0x1d fp=0xc002797b00 sp=0xc002797ae8 pc=0x47f97d
sync.(*Pool).Get(0x0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/.gimme/versions/go1.21.1.linux.amd64/src/sync/pool.go:131 +0x1c fp=0xc002797b38 sp=0xc002797b00 pc=0x47f6fc
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics.getHash({0xc0062c6300, 0x24})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:104 +0x3b fp=0xc002797ba0 sp=0xc002797b38 pc=0x1fbfbbb
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics.RecordEncryptionConfigAutomaticReloadFailure({0xc0062c6300?, 0x40ac040?})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/metrics/metrics.go:86 +0x1c fp=0xc002797be0 sp=0xc002797ba0 pc=0x1fbf91c
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller.(*DynamicKMSEncryptionConfigContent).processNextWorkItem.func1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:184 +0xb3 fp=0xc002797c88 sp=0xc002797be0 pc=0x1fc1a53
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller.(*DynamicKMSEncryptionConfigContent).processNextWorkItem(0xc0051fafc0, {0x52dad98, 0xc0062c82a0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:193 +0x4e7 fp=0xc002797dc0 sp=0xc002797c88 pc=0x1fc1907
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller.(*DynamicKMSEncryptionConfigContent).runWorker(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:147
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller.(*DynamicKMSEncryptionConfigContent).runWorker-fm({0x52dad98, 0xc0062c82a0})
	<autogenerated>:1 +0x49 fp=0xc002797df0 sp=0xc002797dc0 pc=0x1fc21c9
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x22 fp=0xc002797e10 sp=0xc002797df0 pc=0xa97c82
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33 fp=0xc002797e30 sp=0xc002797e10 pc=0xa97b33
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x10?, {0x52a7b00, 0xc004fac030}, 0x1, 0xc0062c82a0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf fp=0xc002797eb0 sp=0xc002797e30 pc=0xa979ef
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0xfa?, 0xc001a1df70?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f fp=0xc002797f00 sp=0xc002797eb0 pc=0xa978df
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x52dad98, 0xc0062c82a0}, 0xc004eed270, 0x3b9aca00?, 0x0?, 0x50?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x93 fp=0xc002797f70 sp=0xc002797f00 pc=0xa97bf3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x52dad98?, 0xc0062c82a0?}, 0x52dad98?, 0xc0062c82a0?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:170 +0x25 fp=0xc002797fb0 sp=0xc002797f70 pc=0xa97825
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller.(*DynamicKMSEncryptionConfigContent).Run.func5()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:87 +0x2d fp=0xc002797fe0 sp=0xc002797fb0 pc=0x1fc096d
runtime.goexit()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/.gimme/versions/go1.21.1.linux.amd64/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc002797fe8 sp=0xc002797fe0 pc=0x473e41
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller.(*DynamicKMSEncryptionConfigContent).Run in goroutine 10620
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go:87 +0x27f
```

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 12]
```diff
@@ -53,6 +53,7 @@ import (
 const (
 	sampleText        = "abcdefghijklmnopqrstuvwxyz"
 	sampleContextText = "0123456789"
+	testAPIServerID   = "testAPIServerID"
```
[the remark this answers is not in the document]
Comment (ritazh) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:56]: it's used in 9 places as variables. might be good to keep it as a const like all other const here?

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:56]: I don't see any test assertions based on the value so I would remove the `const` and just pass the empty string to all the call sites.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 13]
```diff
@@ -111,6 +111,7 @@ func init() {
 	utilruntime.Must(apiserverconfig.AddToScheme(configScheme))
 	utilruntime.Must(apiserverconfigv1.AddToScheme(configScheme))
 	codecs = serializer.NewCodecFactory(configScheme)
+	metrics.RegisterMetrics()
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:114]: You want all 3 metrics packages here to be safe:

```suggestion
	metrics.RegisterMetrics()
	storagevalue.RegisterMetrics()
	encryptionconfigmetrics.RegisterMetrics()
```

Also we probably need to backport this specific change (confirm that these metrics don't work on older releases before doing the backport).

**Review by enj — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go (hunk 13 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:117]: @ritazh opened #120543

**Pull request merged**